### PR TITLE
JS: Simplify helper usage to only one script

### PR DESCRIPTION
**Pull request by LeoColomb**
* Simplify native helper usage and avoid code duplication
  -> Dynamically require specific functions at runtime
* ~~Make it possible to run without Node.js installed (targets #986)
  -> Package the helper into an executable binary with [pkg](https://github.com/zeit/pkg)~~

Note that it is still possible to run the script with Node.js directly, the packaging task only helps to reduce noise.

Also, add back a package-lock file (#897) to allow installing and packing without yarn.

### CODE REVIEW

File: npm_and_yarn/helpers/lib/yarn/index.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const lockfileParser = require("./lockfile-parser");
+const updater = require("./updater");
+const subdependencyUpdater = require("./subdependency-updater");
+const peerDependencyChecker = require("./peer-dependency-checker");
+
+module.exports = {
+  parseLockfile: lockfileParser.parse,
+  update: updater.updateDependencyFiles,
+  updateSubdependency: subdependencyUpdater.updateDependencyFile,
+  checkPeerDependencies: peerDependencyChecker.checkPeerDependencies
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/lib/yarn/index.js]: Would rather make these include the package manager so we can require these explicitly, something like:

```js
module.exports = {
  'yarn:parseLockfile': lockfileParser.parse,
  'yarn:update': updater.updateDependencyFiles,
  'yarn:updateSubdependency': subdependencyUpdater.updateDependencyFile,
  'yarn:checkPeerDependencies': peerDependencyChecker.checkPeerDependencies
}
```

File: npm_and_yarn/helpers/lib/npm/index.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
-const updater = require("../lib/updater");
-const peerDependencyChecker = require("../lib/peer-dependency-checker");
-const subdependencyUpdater = require("../lib/subdependency-updater");
+const updater = require("./updater");
+const peerDependencyChecker = require("./peer-dependency-checker");
+const subdependencyUpdater = require("./subdependency-updater");
 
 module.exports = {
   update: updater.updateDependencyFiles,
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/lib/npm/index.js]: same as yarn:

```
module.exports = {
  'npm:update': updater.updateDependencyFiles,
  'npm:updateSubdependency': subdependencyUpdater.updateDependencyFile,
  'npm:checkPeerDependencies': peerDependencyChecker.checkPeerDependencies
};
```

File: npm_and_yarn/helpers/run.js [hunk 3]
```diff
@@ -1,7 +1,5 @@
 #!/usr/bin/env node
 
-const functionMap = require("../lib");
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/run.js]: Keen to require the helper scripts explicitly here, something like:

```suggestion
const yarnHelpers = require(`./lib/yarn`);
const npmHelpers = require(`./lib/npm`);
const functionMap = { ...yarnHelpers, ...npmHelpers };
```

File: npm_and_yarn/helpers/run.js [hunk 4]
```diff
@@ -10,7 +8,9 @@ const input = [];
 process.stdin.on("data", data => input.push(data));
 process.stdin.on("end", () => {
   const request = JSON.parse(input.join(""));
-  const func = functionMap[request.function];
+  const [manager, functionName] = request.function.split("_");
+  const helpers = require(`./lib/${manager}`);
+  const func = helpers[functionName];
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/run.js]: Could then keep the old functionality here:

```suggestion
  const func = functionMap[request.function];
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher/path_dependency_builder.rb [hunk 5]
```diff
@@ -120,8 +120,8 @@ def parsed_yarn_lock
               File.write("yarn.lock", yarn_lock.content)
 
               SharedHelpers.run_helper_subprocess(
-                command: "node #{yarn_helper_path}",
-                function: "parseLockfile",
+                command: NativeHelpers.helper_path,
+                function: "yarn_parseLockfile",
```
Comment (feelepxyz) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher/path_dependency_builder.rb]: Matching the function to the map

```suggestion
                function: "yarn:parseLockfile",
```

**Comment by LeoColomb** [comment]
Thanks for your review, @feelepxyz!

I've updated the fonction separator to use colon.

Regarding the explicit requires, I'm not against but there is a notable downside: both npm and yarn libs will be loaded at each run, instead of the target one only.

**Comment by feelepxyz** [comment]
@LeoColomb ah yeah good point but think I'm still keen to require them explicitly and change it later if it turns out to be slow. Can't imagine it will add too much overhead, running the scripts is usually the biggest bottleneck.

**Comment by LeoColomb** [comment]
Sure! 🙂 
Did it.

**Comment by feelepxyz** [comment]
@LeoColomb nice one!

Having some issues getting this running on my local dev env using a mac and node 11. Not super familiar with pkg, is it possible to build targets for macos and linux for node 10?

Should also get the instructions in `npm_and_yarn/README.md` to work - looks like the absolute path check is failing here.

**Comment by LeoColomb** [comment]
You may specify node version: 
```shell
$ yarn run build --targets node10
```

See https://github.com/zeit/pkg/blob/master/README.md#targets

I can update the readme, for sure.

**Comment by LeoColomb** [comment]
@feelepxyz I just fixed the README. Should be much better.

**Comment by feelepxyz** [comment]
@LeoColomb thanks for the changes! I've been playing around with this and not sure we want to use `pkg` because it makes local development more painful - having to rebuild the package for changes to apply is slow enough to be a bit annoying 🤠

Would you be up for removing `pkg` and running with node instead?

**Comment by LeoColomb** [comment]
> because it makes local development

I totally understand, that's why I put a note in my initial post that the helper can still be used directly with Node.js without pkg.

I added a commit 18aa49e to use the development option by default and only build when production environment.

Also note pkg usage targets bundled distribution, such as docker image. The main argument for packaging the helper into an executable is to reduced image size and surface (no more full npm lib, no extra Node.js local lib, etc.) and manipulate only one portable binary file.

Let me know if this change solve that situation for you.

**Comment by feelepxyz** [comment]
@LeoColomb cheers! Hm still not sure about `pkg`, not hugely opposed but I don't really see what problem it solves at the moment? Would rather get this in without it and then split up the images as a second step (as you started on) and lastly, if image size is a problem for deploys/startup time, go about tackling that separately. 

Generally keen to keep things simple and not introduce extra complexity if it's not solving a clear problem. Thanks for all your work improving our tooling setup!

**Comment by LeoColomb** [comment]
Alright, reduced to the path refactoring.

File: npm_and_yarn/helpers/build [hunk 6]
```diff
@@ -8,13 +8,22 @@ if [ -z "$install_dir" ]; then
...
+  "$helpers_dir/test" \
+  "$helpers_dir/run.js" \
   "$helpers_dir/package.json" \
   "$helpers_dir/yarn.lock" \
   "$install_dir"
 
 cd "$install_dir"
-yarn install
+
+case "$DEPENDABOT_ENV" in
```
Comment (feelepxyz) [inline npm_and_yarn/helpers/build]: Is this `DEPENDABOT_ENV` used somewhere? Think we should set `NODE_ENV=production` instead when running this script for the same effect.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/native_helpers.rb [hunk 7]
```diff
@@ -3,34 +3,17 @@
...
-
-        File.join(default_helpers_dir, "yarn")
+      def self.native_helpers_root
+        default_path = File.join(__dir__, "../../../helpers/install-dir")
+        ENV.fetch("DEPENDABOT_NATIVE_HELPERS_PATH", default_path)
       end
 
-      def self.default_helpers_dir
-        File.join(__dir__, "../../../../npm_and_yarn/helpers")
+      def self.clean_path(path)
```
Comment (feelepxyz) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/native_helpers.rb]: This now unused?

File: npm_and_yarn/helpers/build (hunk 6 above)
Comment (LeoColomb) [reply npm_and_yarn/helpers/build]: Oops good catch, actually for #986.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/native_helpers.rb (hunk 7 above)
Comment (LeoColomb) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/native_helpers.rb]: Indeed.

**Comment by feelepxyz** [comment]
@LeoColomb sorry for not getting this in, looks like our test suite for npm and yarn went from around 4.5 min to 6.5 min in this PR 😢 will hold off on this for now.

**Comment by LeoColomb** [comment]
> I'm still keen to require them explicitly and change it later if it turns out to be slow.

😇 aff1bd4

**Comment by feelepxyz** [comment]
💯

**Pull request merged**